packages/clerk-js/src/ui/elements/Menu.tsx

```diff
@@ -0,0 +1,175 @@
+import React, { PropsWithChildren, useEffect, useRef } from 'react';
+
+import { createContextAndHook } from '../../ui/utils';
+import { Button, Col } from '../customizables';
+import { usePopover } from '../hooks';
+import { animations, PropsOfComponent } from '../styledSystem';
+import { colors } from '../utils/colors';
+import { Portal } from './Portal';
+
+type UsePopoverReturn = ReturnType<typeof usePopover>;
+
+type MenuState = {
+  popoverCtx: UsePopoverReturn;
+};
+
+const [MenuStateCtx, useMenuState] = createContextAndHook<MenuState>('MenuState');
+
+type MenuProps = PropsWithChildren<Record<never, never>>;
+
+export const Menu = (props: MenuProps) => {
+  const popoverCtx = usePopover({
+    placement: 'right-start',
+    offset: 8,
+  });
+
+  return (
+    <MenuStateCtx.Provider
+      value={{ value: { popoverCtx } }}
+      {...props}
+    />
+  );
+};
+
+type MenuButtonProps = PropsOfComponent<typeof Button>;
+
+export const MenuButton = (props: MenuButtonProps) => {
+  const { popoverCtx } = useMenuState();
+  const { reference, toggle } = popoverCtx;
+
+  return (
+    <Button
+      ref={reference}
+      onClick={toggle}
+      {...props}
+    />
+  );
+};
+
+const findMenuItem = (el: Element, siblingType: 'prev' | 'next', options?: { countSelf?: boolean }) => {
+  let tagName = options?.countSelf ? el.tagName : '';
+  let sibling: Element | null = el;
+  while (sibling && tagName !== 'BUTTON') {
+    sibling = sibling[siblingType === 'prev' ? 'previousElementSibling' : 'nextElementSibling'];
+    tagName = sibling?.tagName ?? '';
+  }
+  return sibling;
+};
+
+type MenuListProps = PropsOfComponent<typeof Col>;
+
+export const MenuList = (props: MenuListProps) => {
+  const { sx, ...rest } = props;
+  const { popoverCtx } = useMenuState();
+  const { floating, styles, isOpen } = popoverCtx;
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    const current = containerRef.current;
+    floating(current);
+    current?.focus();
+  }, [isOpen]);
+
+  const onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+    const current = containerRef.current;
+    if (!current) {
+      return;
+    }
+
+    if (current !== document.activeElement) {
+      return;
+    }
+
+    if (e.key === 'ArrowDown') {
+      return (findMenuItem(current.children[0], 'next', { countSelf: true }) as HTMLElement)?.focus();
```

Comment:
Lot's of things happening here, can we break this statement in multiple lines so its a bit easier to follow along?

Reply:
The `current.children[0]` looks like a low-level way to target the first child and using it makes me _slightly_ uncomfortable.

Can we get the `children` from the props and use that instead?

Reply:
Can you give me an example? The function receives an `Element`.

Reply:
@desiprisg @nikosdouvlis I wonder, couldn't this be simplified using the index of the child as a source of truth? 🙃

example: https://github.com/clerkinc/javascript/commit/08e0eef0dba2ecfbfb3e4ca9663618758161c802

Reply:
@raptisj Do you mean instead of passing the element itself in `findMenuItem`?

Reply:
@desiprisg what you said or a different approach, like passing the index to each child so it is aware of it. Just and idea

```  
const childrenWithProps = getValidChildren(children).map((child, index) =>
    React.cloneElement(child, {
      tabIndex: index,
    }),
  );
```

Reply:
I think if we go that way, it's quite tricky to manage to have no re-renders of the children when navigating with the keyboard. That's why I didn't follow that route.

Reply:
@desiprisg ok got that🙃 

Although since changes are happening in the <Menu />(`isOpen` for example) the children will rerender so you can't really avoid that.

Reply:
Yes, the children will render only when opening or closing the menu, not when navigating through the items with the keyboard. What I'm avoiding in this case, is the re-render of all the children when pressing `ArrowUp` and `ArrowDown`. If the children were aware of their index and I went that route, I'd need to update the context with the newly selected index each time (so the child can check if his index is the selected one), thus causing all the items to re-render.